ludwig/benchmarking/profiler_dataclasses.py

```diff
@@ -0,0 +1,81 @@
+import dataclasses
+from dataclasses import dataclass
+from typing import Dict
+
+from ludwig.utils.data_utils import flatten_dict
+
+
+@dataclass
+class DeviceUsageMetrics:
+    # Max CUDA memory utilization of the code block.
+    max_memory_used: float
+
+    # Average CUDA memory utilization of the code block.
+    average_memory_used: float
+
+
+@dataclass
+class SystemResourceMetrics:
+    # Name of the code block/function to be profiled.
+    code_block_tag: str
+
+    # Name of the CPU that the code ran on.
+    cpu_name: str
+
+    # CPU architecture that the code ran on.
+    cpu_architecture: str
+
+    # Number of CPUs on the machine.
+    num_cpu: int
+
+    # CPU memory available at the beginning of the code block.
+    cpu_memory_available: float
+
+    # Ludwig version in the environment.
+    ludwig_version: str
+
+    # Total execution time of the code block.
+    total_execution_time: float
+
+    # The change in disk memory before and after the code block ran.
+    disk_footprint: float
+
+    # Max CPU utilization of the code block.
+    max_cpu_utilization: float
+
+    # Max CPU memory (RAM) utilization of the code block.
+    max_cpu_memory_usage: float
+
+    # Average CPU utilization of the code block.
+    average_cpu_utilization: float
+
+    # Avergae CPU memory (RAM) utilization of the code block.
+    average_cpu_memory_usage: float
+
+    # Per device usage. Dictionary containing max and average memory used per device.
+    device_usage: Dict[str, DeviceUsageMetrics]
+
+    def to_flat_dict(self):
+        nested_dict = dataclasses.asdict(self)
+        nested_dict[""] = nested_dict.pop("device_usage")
+        return flatten_dict(nested_dict, sep="")
```

Comment:
Pull this out into a separate function, also since it's duplicated across the two classes.

```python
def profiler_dataclass_to_flat_dict(data: Union[SystemResourceMetrics, TorchProfilerMetrics]) -> Dict:
    """Returns a flat dictionary representation, with the device_usage key removed."""
    nested_dict = dataclasses.asdict(data)
    nested_dict[""] = nested_dict.pop("device_usage")
    return flatten_dict(nested_dict, sep="")
```